pySDC/implementations/problem_classes/GenericGusto.py

```diff
@@ -0,0 +1,264 @@
+from pySDC.core.problem import Problem, WorkCounter
+from pySDC.implementations.datatype_classes.firedrake_mesh import firedrake_mesh, IMEX_firedrake_mesh
+from gusto.core.labels import (
+    time_derivative,
+    implicit,
+    explicit,
+)
+from firedrake.fml import replace_subject, all_terms, drop
+import firedrake as fd
+import numpy as np
+
+
+class GenericGusto(Problem):
+    """
+    Set up solvers based on the equation. Keep in mind that you probably want to use the pySDC-Gusto coupling via
+    the `pySDC_integrator` class in the helpers in order to get spatial methods rather than interfacing with this
+    class directly.
+
+    Gusto equations work by a residual, which is minimized in nonlinear solvers to obtain the right hand side
+    evaluation or the solution to (IMEX) Euler steps. You control what you solve for by manipulating labeled parts
+    of the residual.
+    """
+
+    dtype_u = firedrake_mesh
+    dtype_f = firedrake_mesh
+    rhs_n_labels = 1
+
+    def __init__(
+        self,
+        equation,
+        apply_bcs=True,
+        solver_parameters=None,
+        stop_at_divergence=False,
+        LHS_cache_size=12,
+        residual=None,
+        *active_labels,
+    ):
+        """
+        Initialisation
+
+        Args:
+            equation (:class:`PrognosticEquation`): the model's equation.
+            apply_bcs (bool, optional): whether to apply the equation's boundary
+                conditions. Defaults to True.
+            solver_params (dict, optional): Solver parameters for the nonlinear variational problems
+            stop_at_divergence (bool): Whether to raise an error when the variational problems do not converge
+            LHS_cache_size (int): Size of the cache for solvers
```

Comment:
```suggestion
            LHS_cache_size (int, optional): Size of the cache for solvers. Defaults to 12.
```

Reply:
Whoohoo, magic numbers! 😄

Reply:
Well, 12 is a highly composite number and you ideally want something that is divisible by the number of collocation nodes and definitely something that is more than the number of collocation nodes. So it's not as bad as the preconditioners :D